pkg/discovery/dtofactory/pod_entity_dto_builder.go

```diff
@@ -216,6 +241,35 @@ func (builder *podEntityDTOBuilder) getPodCommoditiesSold(pod *api.Pod, cpuFrequ
 	return commoditiesSold, nil
 }
 
+// getPodQuotaCommoditiesSold build the quota commodity DTOs sold by tge pod
+func (builder *podEntityDTOBuilder) getPodQuotaCommoditiesSold(pod *api.Pod, cpuFrequency float64) ([]*proto.CommodityDTO, error) {
+	// CPU resource needs to be converted from number of cores to frequency.
+	converter := NewConverter().Set(func(input float64) float64 {
+		// If CPU resource quota capacity is repository.DEFAULT_METRIC_CAPACITY_VALUE (infinity), skip converting from
+		// cores to frequency.
+		if input == repository.DEFAULT_METRIC_CAPACITY_VALUE {
+			return input
+		}
+		return input * cpuFrequency
+	}, metrics.CPULimitQuota, metrics.CPURequestQuota)
+
+	attributeSetter := NewCommodityAttrSetter()
+	attributeSetter.Add(func(commBuilder *sdkbuilder.CommodityDTOBuilder) { commBuilder.Resizable(false) }, podQuotaCommodities...)
+
+	// Resource Commodities
+	podMId := util.PodMetricIdAPI(pod)
+	quotaCommoditiesSold, err := builder.getResourceCommoditiesSold(metrics.PodType, podMId, podQuotaCommodities, converter, attributeSetter)
+	if err != nil {
+		return nil, err
+	}
+	if len(quotaCommoditiesSold) != len(podQuotaCommodities) {
+		err = fmt.Errorf("mismatch num of bought commidities from node (%d Vs. %d) for pod %s", len(quotaCommoditiesSold), len(podQuotaCommodities), pod.Name)
```

Comment:
nit: Isn't the pod buying its quota commodities from the WorkloadController/Namespace and not the node?

Reply:
Ah sorry I copied this message from the commodity bought from node. I'll fix this and will go through the code again to see if there're other places with such mistake or typo.

And thanks a lot for the suggestion doing a sanity check on market and TP! I tested market results with and without the changes here:
1. No exception or concerning errors in the market component. Container resizing and pod move actions can be normally generated. For container resizing, after the changes, there were no "resize down reservation (request)" actions any more because I removed reservation set on VCPU and VMem, which is fine. And the number of normal resize up/down VCPU/VMem actions matches before and after the changes.
2. No exception in TP but there were 3 different supply chain validation errors like:
```
topology-processor-787849b68d-dg 2020-04-22 22:42:58,859  INFO [realtime-pipeline-runner] [SupplyChainValidator]	: Supply chain validation errors of type class com.vmturbo.topology.processor.supplychain.errors.ProviderCardinalityFailure: 112
topology-processor-787849b68d-dg 2020-04-22 22:42:58,859  INFO [realtime-pipeline-runner] [SupplyChainValidator]	: Supply chain validation errors of type class com.vmturbo.topology.processor.supplychain.errors.MandatoryCommodityBoughtNotFoundFailure: 468
topology-processor-787849b68d-dg 2020-04-22 22:42:58,859  INFO [realtime-pipeline-runner] [SupplyChainValidator]	: Supply chain validation errors of type class com.vmturbo.topology.processor.supplychain.errors.MandatoryCommodityNotFoundFailure: 135
```
which are `ProviderCardinalityFailure`, `MandatoryCommodityBoughtNotFoundFailure` and `MandatoryCommodityNotFoundFailure`

I think I know the reason why these are failed. 

For example, for "ProviderCardinalityFailure", the pod supply chain node is built in `supply_chain_factory.go` by:
```
podSupplyChainNodeBuilder.
    Provider(proto.EntityDTO_VIRTUAL_MACHINE, proto.Provider_HOSTING).
    Buys(...).
    Provider(proto.EntityDTO_WORKLOAD_CONTROLLER, proto.Provider_HOSTING).
    Buys(...).
    Provider(proto.EntityDTO_NAMESPACE, proto.Provider_HOSTING).
    Buys(...).
    ...
```
which means there are 3 mandatory hosts for the pod, but our use case is that Pod is hosted by the node, and could be hosted by either WorkloadController or Namespace.

Then for "MandatoryCommodityBoughtNotFoundFailure", the container supply chain node is built by:
```
Provider(proto.EntityDTO_CONTAINER_POD, proto.Provider_HOSTING).
		...
		Buys(vCpuRequestTemplateComm).
		Buys(vMemRequestTemplateComm).
                ...
```
Looks in this way, it's mandatory for container to buy VCPURequest and VMemRequest. But in the changes here, if a container doesn't have requests set, then it won't buy requests commodities, which leads to supply chain failure. The same case for `MandatoryCommodityNotFoundFailure` for the requests commodities sold by Container.

Do you happen to know if there's a way to set **optional** host and **optional** TemplateCommodity in supply chain? I probably need to reach out to Yannis for help tomorrow.

Reply:
Additional changes are needed to fix the supply chain failure in turbo-go-sdk and kubeturbo. Will put those in a following PR to avoid making this PR hard to follow.